include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   */
+  virtual WasmResult setTimerPeriod(std::chrono::milliseconds /* period */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Provides the current time in nanoseconds.
+   */
+  virtual uint64_t getCurrentTimeNanoseconds() {
+    struct timespec tpe;
+    clock_gettime(CLOCK_REALTIME, &tpe);
+    uint64_t t = tpe.tv_sec;
+    t *= 1000000000;
+    t += tpe.tv_nsec;
+    return t;
+  }
+
+  /**
+   * Provides the status of the call into the VM or out of the VM.
+   * @return the status code and a descriptive string.
+   */
+  virtual std::pair<uint32_t, string_view> getStatus() {
+    unimplemented();
+    return std::make_pair(1, "unimplmemented");
+  }
+
+  // Buffer
+
+  /**
+   * Provides a BufferInterface to be used to return buffered data to the VM.
+   * @param type is the type of buffer to provide.
+   */
+  virtual const BufferInterface *getBuffer(WasmBufferType /* type */) {
+    unimplemented();
+    return nullptr;
+  }
+
+  /**
+   * Provides the end-of-stream status of the current stream (if any) or false.
+   */
+  virtual bool end_of_stream() {
+    unimplemented();
+    return true;
+  }
+
+  // HTTP
+
+  /**
+   * Make an outgoing HTTP request.
+   * @param target specifies the proxy-specific target of the call (e.g. a cluster, peer, or host).
+   * @param request_headers are the request headers.
+   * @param request_body is the request body.
+   * @param request_trailers are the request trailers.
+   * @param timeout_milliseconds is a timeout after which the request will be conidered to have
+   * failed.
+   * @param token_ptr contains a pointer to a location to store the token which will be used with
+   * the corresponding onHttpCallResponse.
+   */
+  virtual WasmResult httpCall(string_view /* target */, const Pairs & /*request_headers */,
+                              string_view /* request_body */, const Pairs & /* request_trailers */,
+                              int /* timeout_millisconds */, uint32_t * /* token_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // gRPC
+
+  /**
+   * Make a gRPC call.
+   * @param grpc_service is proxy-specific metadata describing the service (e.g. security certs).
+   * @param service_name the name of the gRPC service.
+   * @param method_name the gRPC method name.
+   * @param request the serialized request.
+   * @param initial_metadata the initial metadata.
+   * @param timeout a timeout in milliseconds.
+   * @param token_ptr contains a pointer to a location to store the token which will be used with
+   * the corresponding onGrpc and grpc calls.
+   */
+  virtual WasmResult grpcCall(string_view /* grpc_service */, string_view /* service_name */,
+                              string_view /* method_name */, string_view /* request */,
+                              const Pairs & /* initial_metadata */,
+                              const optional<std::chrono::milliseconds> & /* timeout */,
```

Comment:
I would avoid having an optional timeout for this type of API

Reply:
You want it to build part of the grpc_service configuration?

Reply:
I just meant that we should require a timeout. The reason it is this way in Envoy is I think for historical reasons. I think here we should just require the user to set a timeout as a good practice?

Reply:
AH, got it.  I agree.   What does it mean when it is not present ?

Reply:
I suppose that if we don't provide the timeout currently that there is no timeout, which is the same as just providing a sufficiently large one. 

Done!